keps/sig-storage/1698-generic-ephemeral-volumes/README.md

```diff
@@ -0,0 +1,566 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes an alternative mechanism for specifying volumes
+inside a pod spec. Those inline volumes then get converted to normal
+volume claim objects for provisioning, so storage drivers do not need
+to be modified.
+
+The lifecycle of those volumes is ephemeral: they will get deleted
+once the pod terminates.
+
+Because this is expected to be slower than [CSI ephemeral inline
+volumes](https://github.com/kubernetes/enhancements/issues/596), both
+approaches will need to be supported.
+
+
+## Motivation
+
+The current CSI ephemeral inline volume feature has demonstrated that
+ephemeral inline volumes are a useful concept, for example to provide
+additional scratch space for a pod. Several CSI drivers support it
+now, most of them in addition to normal volumes (see
+https://kubernetes-csi.github.io/docs/drivers.html).
+
+However, the original design was intentionally focused on
+light-weight, local volumes. It is not a good fit for volumes provided
+by a more traditional storage system because:
+- The normal API for selecting volume parameters (like size and
+  storage class) is not supported.
+- Integration into storage capacity aware pod scheduling is
+  challenging and would depend on extending the CSI ephemeral inline
+  volume API.
+- CSI drivers need to be adapted and have to take over some of the
+  work normally done by Kubernetes, like tracking of orphaned volumes.
+
+
+### Goals
+
+- Volumes can be specified inside the pod spec ("inline").
+- Volumes are created for specific pods and deleted after the pod
+  terminates ("ephemeral").
+- A normal, unmodified storage driver can be selected via a storage class.
+- The volume will be created using the normal storage provisioning
+  mechanism, without having to modify the driver or its deployment.
+- Storage capacity tracking can be enabled also for such volumes.
+
+### Non-Goals
+
+- This will not replace CSI ephemeral inline volumes because the
+  goals for those (light-weight local volumes) are not a good fit
+  for the approach proposed here.
+- These inline volumes will always be ephemeral. If making them persistent
+  was allowed, some additional controller would be needed to manage them
+  after pod termination, in which case it will probably be simpler to
+  also create them separately.
+
+## Proposal
+
+A new volume source will be introduced:
+
+```
+type EphemeralVolumeSource struct {
+    // Will be created as a stand-alone PVC to provision the volume.
+    // Required, must not be nil.
+    VolumeClaimTemplate *PersistentVolumeClaim
+    ReadOnly             bool
```

Comment:
We should think twice about if this ReadOnly flag is really necessary (https://github.com/kubernetes/kubernetes/issues/70503). The answer may still be yes because it simplifies implementation, but if we can avoid making the problem worse, that would be great.